hollow/src/main/java/com/netflix/hollow/core/memory/ByteArrayOrdinalMap.java

```diff
@@ -43,471 +25,554 @@
+                }
+
+                populatedReverseKeys[i] = populatedReverseKeys[i] << BITS_PER_POINTER | currentCopyPointer;
+
+                currentCopyPointer += length;
+            } else {
+                freeOrdinalTracker.returnOrdinalToPool(ordinal);
+                populatedReverseKeys[i] = EMPTY_BUCKET_VALUE;
+            }
+        }
+
+        byteData.setPosition(currentCopyPointer);
+        freeOrdinalTracker.sort();
+
+        // Reset the array then fill with compacted values
+        // Volatile store not required, could use plain store
+        // See VarHandles for JDK >= 9
+        for (int i = 0; i < pao.length(); i++) {
+            pao.lazySet(i, EMPTY_BUCKET_VALUE);
+        }
+        populateNewHashArray(pao, populatedReverseKeys);
+        size = usedOrdinals.cardinality();
+
+        pointersByOrdinal = null;
+        unusedPreviousOrdinals = null;
+    }
+
+    public long getPointerForData(int ordinal) {
+        long pointer = pointersByOrdinal[ordinal] & POINTER_MASK;
+        return pointer + VarInt.nextVLongSize(byteData.getUnderlyingArray(), pointer);
+    }
+
+    public boolean isReadyForWriting() {
+        return pointersByOrdinal != null;
+    }
+
+    public boolean isReadyForAddingObjects() {
+        return pointersByOrdinal == null;
+    }
+
+    public long getDataSize() {
+        return byteData.length();
+    }
+
+    public int maxOrdinal() {
+        int maxOrdinal = -1;
+        AtomicLongArray pao = pointersAndOrdinals;
+
+        for (int i = 0; i < pao.length(); i++) {
+            long key = pao.get(i);
+            if (key != EMPTY_BUCKET_VALUE) {
+                int ordinal = (int) (key >>> BITS_PER_POINTER);
+                if (ordinal > maxOrdinal) {
+                    maxOrdinal = ordinal;
+                }
+            }
+        }
+        return maxOrdinal;
+    }
+
+    /**
+     * Compare the byte sequence contained in the supplied ByteDataBuffer with the
+     * sequence contained in the map pointed to by the specified key, byte by byte.
+     */
+    private boolean compare(ByteDataBuffer serializedRepresentation, long key) {
+        long position = key & POINTER_MASK;
+
+        int sizeOfData = VarInt.readVInt(byteData.getUnderlyingArray(), position);
+
+        if (sizeOfData != serializedRepresentation.length()) {
+            return false;
+        }
+
+        position += VarInt.sizeOfVInt(sizeOfData);
+
+        for (int i = 0; i < sizeOfData; i++) {
+            if (serializedRepresentation.get(i) != byteData.get(position++)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Resize the ordinal map by increasing its capacity.
+     * <p>
+     * No action is take if the current capacity is sufficient for the given size.
+     *
+     * @param size the size to increase to, rounded up to the nearest power of two.
+     */
+    public void resize(int size) {
```

Comment:
May want to add an admonition similar to the one above `put()` that this operation is not thread-safe (since it can result in missed updates from the calls to `getOrAssignOrdinal`).  This scenario would also leave the `FreeOrdinalTracker` out of sync.

Reply:
I updated the method documentation.

IIUC since a resize does not affect ordinal assignment (just the redistribution in the map) it means the free ordinal tracker size is out of sync, which previously would occur as ordinals get assigned (since the tracker size increases as ordinals are returned to the pool).

Looks like there is some opportunity to better size the tracker in `recalculateFreeOrdinals` (length - cardinality), although we could avoid the allocation of a temporary bit set instead. I am less certain about compaction.